lib/src/rewrite.rs

```diff
@@ -464,70 +464,107 @@ pub struct MoveCommitsStats {
+                        RevsetEvaluationError::StoreError(err) => err,
+                        RevsetEvaluationError::Other(_) => panic!("Unexpected revset error: {err}"),
+                    })?
+                    .iter()
+                    .commits(mut_repo.store())
+                    .try_collect()?;
+            connected_target_commits_internal_parents =
+                compute_connected_target_commits_internal_parents(
+                    &target_commit_ids,
+                    &connected_target_commits,
+                );
+
+            target_roots = connected_target_commits_internal_parents
+                .iter()
+                .filter(|(commit_id, parents)| {
+                    target_commit_ids.contains(commit_id) && parents.is_empty()
+                })
+                .map(|(commit_id, _)| commit_id.clone())
+                .collect();
+        }
+        MoveCommitsTarget::Roots(roots) => {
+            if roots.is_empty() {
+                return Ok(MoveCommitsStats {
+                    num_rebased_targets: 0,
+                    num_rebased_descendants: 0,
+                    num_skipped_rebases: 0,
+                    num_abandoned: 0,
+                });
+            }
 
-    // Compute the parents of all commits in the connected target set, allowing only
-    // commits in the target set as parents.
-    let connected_target_commits_internal_parents =
-        compute_connected_target_commits_internal_parents(
-            &target_commit_ids,
-            &connected_target_commits,
-        );
-
-    // Compute the roots of `target_commits` if not provided.
-    let target_roots: HashSet<_> = if let Some(target_roots) = target_roots {
-        target_roots.iter().cloned().collect()
-    } else {
-        connected_target_commits_internal_parents
-            .iter()
-            .filter(|(commit_id, parents)| {
-                target_commit_ids.contains(commit_id) && parents.is_empty()
-            })
-            .map(|(commit_id, _)| commit_id.clone())
-            .collect()
-    };
+            target_commits = RevsetExpression::commits(roots.iter().ids().cloned().collect_vec())
+                .descendants()
+                .evaluate_programmatic(mut_repo)
+                .map_err(|err| match err {
+                    RevsetEvaluationError::StoreError(err) => err,
+                    RevsetEvaluationError::Other(_) => panic!("Unexpected revset error: {err}"),
+                })?
+                .iter()
+                .commits(mut_repo.store())
+                .try_collect()?;
+            target_commit_ids = target_commits.iter().ids().cloned().collect();
+
+            connected_target_commits = target_commits.iter().cloned().collect_vec();
+            connected_target_commits_internal_parents =
+                compute_connected_target_commits_internal_parents(
+                    &target_commit_ids,
+                    &connected_target_commits,
+                );
```

Comment:
We don't have to resolve internal parents because targets are identical to connected targets?